Add draft report on AI agents (#34)
w3c/ai-web-impact: agent-report.md

Some of these sources have already well-defined entry points: the [webref](https://github.com/w3c/webref/) project provides a comprehensive repository of machine-readable data on browser technologies, the [web features project](https://github.com/web-platform-dx/web-features/) catalogs a description of all browser features which the [browser-compat-data](https://github.com/mdn/browser-compat-data/) allows to associate with actual shipping implementations; the [MDN documentation project](https://github.com/mdn/content/) curates technical reference documentation and training materials on these features, with a specific focus on accuracy, and integrating qualitative considerations about their usage (e.g. to ensure their accessible use).

Exploring whether these sources should be packaged and distributed in a way that would increase their impact on existing and emerging AI Agents may help increase quality adoption of new web technologies. It may also pave the way for a broader conversation on how technology ecosystems in general can best integrate with coding AI agents.

Risks:

* 

Opportunities:

> * 

Reviewer: ```suggestion
* Coding AI assistants may make web development much more accessible
```

> # AI Agents and the Developer Experience
> 
> One domain where AI Agents seem to be gaining rapid adoption is in coding environments. Since the impact of most Web technologies ultimately depends on how well and rapidly Web developers can make use of them, it is useful to consider what impact coding AI Agents might have on how W3C's work gets as well received as possible in these new circumstances.
> 
> Today, the following elements are parts of the adoption path for developers:
> 
> * Awareness that the said technology or feature exists,
> * Availability and quality of implementations, from experimental to shipped interoperably (as recognized e.g. by "[Baseline, widely available](https://web-platform-dx.github.io/web-features/)" label for Web features)  
> * Availability of documentation and training material,  
> * Availability of coding support (e.g. IDE autocompletion, templates, refactoring),  
> * Availability of supporting libraries and frameworks,  
> * Ease of integration into existing systems (e.g. a payment technology needs to work with existing payment workflows).
> 
> Coding AI Agents can provide increased visibility and support in almost any of these elements:
> 
> * Prompts can help surface what specific technologies and features are relevant to accomplish a task, or AI Agents can provide direct code that make use of these.  
> * AI Agents can propose or annotate code based on information on available implementations informed by the expected deployment target.  
> * AI Agents can summarize or surface relevant technical documentation in context, and can help structure, adapt and run a  training program on a specific technology,  
> * AI Agents already provide supercharged capabilities in terms of code completion, templates and refactoring, now extending to extensive code production,  
> * AI Agents could presumably help identify libraries and frameworks that help adopt a given feature, although their ability to do that at scale and without hallucinations remains uncertain,  
> * Likewise, AI Agents can provide specific guidance to integrate a given feature within an existing workflow or system based on the patterns it might have identified in its training set.
> 
> To fully realize these opportunities, it is critical to ensure that these AI Agents have access to **accurate** and **up-to-date** information about the said features, libraries and systems.
> 
> Because that information tends to be fast evolving, relying on up-to-date material being incorporated in the training phase of the underlying models is unlikely to yield satisfactory results. It is much more likely that providing well-identified sources of quality information for Web technologies that agents could integrate using e.g. MCP would ensure a much better outcome.
> 
> Some of these sources have already well-defined entry points: the [webref](https://github.com/w3c/webref/) project provides a comprehensive repository of machine-readable data on browser technologies, the [web features project](https://github.com/web-platform-dx/web-features/) catalogs a description of all browser features which the [browser-compat-data](https://github.com/mdn/browser-compat-data/) allows to associate with actual shipping implementations; the [MDN documentation project](https://github.com/mdn/content/) curates technical reference documentation and training materials on these features, with a specific focus on accuracy, and integrating qualitative considerations about their usage (e.g. to ensure their accessible use).
> 
> Exploring whether these sources should be packaged and distributed in a way that would increase their impact on existing and emerging AI Agents may help increase quality adoption of new web technologies. It may also pave the way for a broader conversation on how technology ecosystems in general can best integrate with coding AI agents.

Reviewer 2: ```suggestion
# AI Agents and the Developer Experience

A [large developer survey in 2024](https://survey.stackoverflow.co/2024/) confirmed that a majority of developers now use AI tools to assist in coding, for example through code auto-completion, templates and refactoring, and extensive code production. We expect AI Agents will soon be able to:

* Propose specific technologies or features (or even code) to accomplish a given task.
* Propose or annotate code based on information on available implementations and the expected deployment target.  
* Summarize technical documentation tailored for a give context, or create a training program.
* Identify libraries and frameworks that help adopt a given feature (although their ability to do that at scale and without hallucinations remains uncertain).
* Guide feature integration within an existing workflow or system based on the patterns it might have identified in its training set.

As with any other domain, the utility and quality of AI tools for coding with Web technology depends on accurate, up-to-date information. Because the Web is large, and constantly evolving without centralized control, a training strategy that relies indiscriminately on code readily found on the Web (which may be out-of-date) may not yield good results. Instead, integrating curated information (e.g., using Model Context Protocol (MCP)) would ensure better outcomes.

Today's curated resources include:

* the [webref](https://github.com/w3c/webref/) project, a comprehensive repository of machine-readable data on browser technologies
* the [web features project](https://github.com/web-platform-dx/web-features/), catalog of descriptions of all browser features which the [browser-compat-data](https://github.com/mdn/browser-compat-data/) allows to associate with actual shipping implementations
* the [MDN documentation project](https://github.com/mdn/content/), curated technical reference documentation and training materials on these features, with a specific focus on accuracy, and integrating qualitative considerations about their usage (e.g., to support accessibility)

Packaging and distributing these resources with greater focus on their impact on existing and emerging AI Agents may help increase quality adoption of new web technologies. It may also pave the way for a broader conversation on how to integrate AI agents into other technology ecosystems.
```

Author: thanks! I've adopted most (but not all) of these changes in c4461a0

# AI Agents Interacting with the Web

The Web is arguably the primary platform to expose digital content and services and constitutes thus an unavoidable source and target for AI Agents.  
We posit that the impact of AI Agents will differ vastly depending on the nature of Web end points: in particular, we describe below how the impact may differ for *content* compared to *online services*.

## Impact on Web Content

AI Agents are likely to have a significant impact on common Web business models. 

Many Web sites rely on users directly perceiving content for monetization (e.g. via advertising). At the same time, many sites rely on analytics in order to tailor content to users, gather insights for new content, and justify the expense of producing the content. 

AI Agents are likely to have an impact on both monetization and analytics, but the impact is likely to vary based on two main usage scenarios:

* Targeted consumption (e.g,. watching a movie, listening to music),  
* Information gathering (e.g., picking a vacation destination). 

> For the latter user experiences, so-called "deep research"  AI agents have already appeared on the scene. They aim to replace some types of search and browsing across the Web, and have the effect of limiting or even eliminating the user’s interactions with a multiplicity of sites. By doing so, AI Agents directly impact monetization opportunities for those sites, which may jeopardize their livelihood. Early signals of reduced visits from search engines suggest this is already occurring. Unless alternatives are established to mitigate this impact, a significant portion of content producers may no longer be able to distribute their content openly, reducing the overall value of the Web both for end users and AI Agents.

Reviewer 2: ```suggestion
For the latter user experiences, so-called "deep research" AI agents have already appeared on the scene. Some operate on traditional search sites, but others move search capabilities directly into the browser and may replace some types of traditional search engine web sites. In doing so, these AI Agents directly impact monetization opportunities for those sites, which may jeopardize their livelihood. Early signals of reduced visits from search engines suggest this is already occurring. Unless alternatives are established to mitigate this impact, a significant portion of content producers may no longer be able to distribute their content openly, reducing the overall value of the Web both for end users and AI Agents.
```

In such an ecosystem, or already today for content providers with no or limited dependency on monetization or analytics, another need is emerging to make AI Agents more effective in their tasks: customizing how content is delivered or made discoverable to them ([llms.txt](https://llmstxt.org/), exposure of data and APIs through MCP). 

Similar situations have arisen at different points of the Web history: the development of content tailored specifically for people with disabilities, or for search engines, or for mobile devices. While there are indeed situations where these alternative views are justified or even required, the value of maintaining the [one Web principle](https://www.w3.org/TR/ethical-web-principles/#oneweb) points towards a path facilitating convergence with existing discovery and rendering mechanisms rather than ones that strengthen separate windows on the Web. In particular, it seems likely in this case as in previous ones that a semantic-rich progressive-enhancement based approach to developing content would provide a strong basis to serve the needs from AI Agents.

Risks:

* Depletion of openly distributed Web content  

> * New intermediation between content producers and their users creates a disconnect in the feedback loop  

Author: ```suggestion
* New intermediation between content producers and their users creates a disconnect in the feedback loop  and reduce brand recognition for content creators
```

Reviewer 3: Looks good, thank you.

AI Agents will themselves be working together, so another important brick involves **service orchestration**. Long ago W3C explored standardization in this area through the [Web Services stack](https://www.w3.org/TR/ws-arch). More recently and more relevant, W3C's evolving [Web of Things stack](https://www.w3.org/TR/wot-architecture11/) can be used in this environment of agent to agent communication, as illustrated by the [Eclipse LMOS protocol](https://eclipse.dev/lmos/docs/lmos_protocol/introduction/).   
There has been at least one significant change in circumstances since these stacks were originally developed: LLMs are now able to translate concepts across formats and languages. Even in this new world, we can learn lessons from previous efforts, such as analyzing why Web Services, though strongly supported by key industry players, failed to have their anticipated impact.

Risks:

* 

Opportunities:

* 

# AI Agents as Web User Agents

As discussed above, AI Agents can be used to replace what end users would typically have used their browsers for when researching and gathering information online.

> This points towards a convergence that is actively being explored across several projects and organizations: AI Agents as complements or replacements to browsers as primary web user agents.

Reviewer 3: ```suggestion
This points towards a convergence that is actively being explored across several projects and organizations: AI Agents as complements or replacements for browsers as primary web user agents.
```
```suggestion
This points towards a convergence that is actively being explored across several projects and organizations: AI Agents as complements or replacements to browsers as primary web user agents.
```

From an end user perspective, the primary value of AI Agents is likely to reside in their ability to remove or reduce friction, i.e. to automate tasks users are currently required to perform but do not perceive to benefit from.

Browsers as they currently exist already compete on providing some of that support in their user interface: authentication managers, smart autocomplete, payment, malware protection are all illustration of automation or semi-automation that have been deployed to make users life easier.

AI Agents bring a number of potential significant additional improvements given their ability to interpret natural language commands from users, natural language used on Web pages, and to enact complex workflows across pages and sites (incl. form filling and submission).

To take full advantage of these benefits, the following structural issues need to be addressed:

> * LLM-based systems suffer from so-called "hallucinations" that may lead AI Agents to take nonsensical or harmful actions; depending on the sensitivity activity they would be automating, these hallucinations could create significant harm to the end users they're supposed to represent, with significant detrimental impact on the trust end users would have in agents, but also possibly in online systems in general.  

Reviewer 3: ```suggestion
* LLM-based systems suffer from so-called "hallucinations" that may lead AI Agents to take nonsensical or harmful actions; depending on the sensitivity of the activity they would be automating, these hallucinations could create significant harm to the end users they're supposed to represent, with significant detrimental impact on the trust end users would have in agents, but also possibly in online systems in general.  
```
```suggestion
* LLM-based systems suffer from so-called "hallucinations" that may lead AI Agents to take nonsensical or harmful actions; depending on the sensitivity activity they would be automating, these hallucinations could create significant harm to the end users they're supposed to represent, with significant detrimental impact on the trust end users would have in agents, but also possibly in online systems in general.  
```

* Over time, web browsers have adopted (more or less robustly) a series of duties they are expected to follow, as the W3C Technical Architecture Group is in the [process of documenting](https://w3ctag.github.io/user-agents/). Beyond any moral or legal considerations, these duties have played a critical role in enabling a flourishing ecosystem for both end users and content providers. The extent to which these duties can be applied in the less controlled environment induced by LLM-powered systems remains to be determined; for instance, browsers take great pain in avoiding to expose what sites a user has already visited \- are there risks that an AI-based system would risk surfacing not only browsing history but maybe even prompt history?  

> Generally speaking, the current user experience of web browsers provide a relatively clear liability story, where the end user, the browser and the content can be assigned blame when something goes wrong. In a world where an AI agent takes more of the end user tasks and operates through a less deterministic model than a web browser, possibly to interact with content itself managed by an AI agent with its own limitations, the risk of exposing users to an unreliable and untrustworthy platform is significant.

Reviewer 3: ```suggestion
Generally speaking, the current user experience of web browsers provides a relatively clear liability story, where the end user, the browser and the content can be assigned blame when something goes wrong. In a world where an AI agent takes more of the end user tasks and operates through a less deterministic model than a web browser, possibly to interact with content itself managed by an AI agent with its own limitations, the risk of exposing users to an unreliable and untrustworthy platform is significant.
```
```suggestion
Generally speaking, the current user experience of web browsers provide a relatively clear liability story, where the end user, the browser and the content can be assigned blame when something goes wrong. In a world where an AI agent takes more of the end user tasks and operates through a less deterministic model than a web browser, possibly to interact with content itself managed by an AI agent with its own limitations, the risk of exposing users to an unreliable and untrustworthy platform is significant.
```

Another assumption that these AI web user agents have already started breaking is the relatively clean delineation between end-user operated agents and crawlers. When crawlers emerged in the early days of the Web, the impact of their automated web page access on web servers led to the creation of the robots.txt proposal indicating which pages to crawl and limiting the frequency of access. In practice, there were only so many crawlers deployed at web scale that this proved sufficient in keeping the induced load on servers manageable.

AI agents conducting e.g. deep research tasks for the end users can fetch many pages in quick succession in a completely distributed fashion across the population of end users \- as for instance a number of open source project documentation web sites have recently experienced, to their detriment.  
Where robots.txt was meant to manage the impact of a few non-human directed crawlers, it is unlikely to be fit for purpose for widely distributed, user-directed agents, all the less so as they're unlikely to be recognizable as "robot" user agents given their nature as new generation of  browser.  

> This lack of clarity has made a number of anti-AI firewalls approaches emerge, which create friction and expand more resources \- clarifying the expected behaviors of this new type of content fetchers would not solve misbehaving actors, but would likely go a long way to reduce the overall cost on the ecosystem.  

Reviewer 2: Authentication managers, smart autocomplete, payments, and malware protection all illustrate how browser automation or semi-automation can make our lives on the Web a bit easier. In some cases, these capabilities blur the line between "the browser" and "the Web." For example, in most browsers the "URL bar" is no longer just for URLs: users can type queries that are then fed to a search engine running on a Web site. This convenient UX leads many users to think that the user agent is performing the search, even though a Web site did so. As mentioned above, AI Agents will actually move some search capabilities into the browser, further blurring the browser/Web distinction when it comes to search. New automation enhancements such as enabling users to issue natural language commands to the browser or web site, and enacting complex workflows across pages and sites (incl. form filling and submission) will further blur the line, and more importantly, raise important questions of trust.

For example, LLM-based systems suffer from so-called "hallucinations" that may lead AI Agents to take nonsensical or harmful actions; depending on the sensitivity activity they would be automating, these hallucinations could create significant harm to the end users they're supposed to represent, with significant detrimental impact on the trust end users would have in agents, but also possibly in online systems in general. The inability to establish a robust browser/Web distinction in LLM-based systems has already allowed prompt-injection attacks in many contexts. In an environment as open as the Web, and from a context as sensitive as the one managed today by browsers, such attacks could cause serious harm to users (see also AI in the Browser Threat Modeling).

A longstanding principle of Web architecture has been that the browser represents the user's interests. As the W3C Technical Architecture Group (TAG) writes, "A user agent acts as an intermediary between a person (its user) and the web... The user agent serves the person as a trustworthy agent: it always puts that person's interest first." As AI Agents transform the browser landscape, how will they earn our trust?

Browsers have (to a large extent) earned our user trust by performing "duties" on behalf of the user. Beyond any moral or legal considerations, these duties have played a critical role in enabling a flourishing ecosystem for both end users and content providers. The extent to which these duties can be applied in the less controlled environment induced by LLM-powered systems remains to be seen. For instance, browsers take pains to not expose the sites a user has already visited. Are there risks that an AI-based system would risk surfacing not only browsing history but maybe even prompt history?

Generally speaking, the current user experience of web browsers tells a relatively clear liability story: when something goes wrong we can tell whether there was user error, browser software error, or the problem lay with the Web site. When an AI agent takes more of the end user tasks and operates through a less deterministic model than a web browser, there is a significant risk of exposing users to an unreliable and untrustworthy platform. The risk will likely be compounded as the user's AI agents interact with a Web site's AI agents.

Another assumption that AI web user agents are proving fragile is the relatively clean delineation between end-user operated agents and crawlers. When crawlers emerged in the early days of the Web, the impact of their automated web page access on web servers led to the creation of the robots.txt proposal indicating which pages to crawl and limiting the frequency of access. In practice, there were only so many crawlers deployed at web scale that this proved sufficient in keeping the induced load on servers manageable. AI agents conducting deep research tasks for the end users can fetch many pages in quick succession in a completely distributed fashion across the population of end users (as painfully felt by a number of open source projects). Where robots.txt was meant to manage the impact of a few non-human directed crawlers, it is unlikely to be fit for purpose for widely distributed, user-directed agents (which may not be easily recognizable as traditional crawlers). Without a standard in place for managing these interactions, anti-AI firewalls have started to emerge, which creates friction and requires more resources. We see standardization opportunities to support responsible actors that wish to work together to reduce ecosystem costs.

> As discussed above, AI Agents can be used to replace what end users would typically have used their browsers for when researching and gathering information online.
> 
> This points towards a convergence that is actively being explored across several projects and organizations: AI Agents as complements or replacements to browsers as primary web user agents.
> 
> From an end user perspective, the primary value of AI Agents is likely to reside in their ability to remove or reduce friction, i.e. to automate tasks users are currently required to perform but do not perceive to benefit from.
> 
> Browsers as they currently exist already compete on providing some of that support in their user interface: authentication managers, smart autocomplete, payment, malware protection are all illustration of automation or semi-automation that have been deployed to make users life easier.
> 
> AI Agents bring a number of potential significant additional improvements given their ability to interpret natural language commands from users, natural language used on Web pages, and to enact complex workflows across pages and sites (incl. form filling and submission).
> 
> To take full advantage of these benefits, the following structural issues need to be addressed:
> 
> * LLM-based systems suffer from so-called "hallucinations" that may lead AI Agents to take nonsensical or harmful actions; depending on the sensitivity activity they would be automating, these hallucinations could create significant harm to the end users they're supposed to represent, with significant detrimental impact on the trust end users would have in agents, but also possibly in online systems in general.  
> * A critical architectural design of Web browsers to protect their end users is the sharp separation between browser-controlled UI (the so-called "chrome") and the interface created and managed by Web content. The inability to establish as robust a separation in LLM-based systems has already allowed prompt-injection attacks in many contexts; in an environment as open as the Web in a context as sensitive as the one controlled by browsers, the impact of these attacks would be potentially manyfold worse (see also [AI in the Browser Threat Modelling](https://github.com/w3c-cg/threat-modeling/blob/simoneonofri-patch-2/models/ai-in-browser.md)).  
> * Over time, web browsers have adopted (more or less robustly) a series of duties they are expected to follow, as the W3C Technical Architecture Group is in the [process of documenting](https://w3ctag.github.io/user-agents/). Beyond any moral or legal considerations, these duties have played a critical role in enabling a flourishing ecosystem for both end users and content providers. The extent to which these duties can be applied in the less controlled environment induced by LLM-powered systems remains to be determined; for instance, browsers take great pain in avoiding to expose what sites a user has already visited \- are there risks that an AI-based system would risk surfacing not only browsing history but maybe even prompt history?  
> 
> Generally speaking, the current user experience of web browsers provide a relatively clear liability story, where the end user, the browser and the content can be assigned blame when something goes wrong. In a world where an AI agent takes more of the end user tasks and operates through a less deterministic model than a web browser, possibly to interact with content itself managed by an AI agent with its own limitations, the risk of exposing users to an unreliable and untrustworthy platform is significant.
> 
> Another assumption that these AI web user agents have already started breaking is the relatively clean delineation between end-user operated agents and crawlers. When crawlers emerged in the early days of the Web, the impact of their automated web page access on web servers led to the creation of the robots.txt proposal indicating which pages to crawl and limiting the frequency of access. In practice, there were only so many crawlers deployed at web scale that this proved sufficient in keeping the induced load on servers manageable.
> 
> AI agents conducting e.g. deep research tasks for the end users can fetch many pages in quick succession in a completely distributed fashion across the population of end users \- as for instance a number of open source project documentation web sites have recently experienced, to their detriment.  
> Where robots.txt was meant to manage the impact of a few non-human directed crawlers, it is unlikely to be fit for purpose for widely distributed, user-directed agents, all the less so as they're unlikely to be recognizable as "robot" user agents given their nature as new generation of  browser.  
> This lack of clarity has made a number of anti-AI firewalls approaches emerge, which create friction and expand more resources \- clarifying the expected behaviors of this new type of content fetchers would not solve misbehaving actors, but would likely go a long way to reduce the overall cost on the ecosystem.  

Reviewer 2: ```suggestion
Authentication managers, smart autocomplete, payments, and malware protection all illustrate how browser automation or semi-automation can make our lives on the Web a bit easier. In some cases, these capabilities blur the line between "the browser" and "the Web." For example, in most browsers the "URL bar" is no longer just for URLs: users can type queries that are then fed to a search engine running on a Web site. This convenient UX leads many users to think that the user agent is performing the search, even though a Web site did so. As mentioned above, AI Agents will actually move some search capabilities into the browser, further blurring the browser/Web distinction when it comes to search. New automation enhancements such as enabling users to issue natural language commands to the browser or web site, and enacting complex workflows across pages and sites (incl. form filling and submission) will further blur the line, and more importantly, raise important questions of trust.

For example, LLM-based systems suffer from so-called "hallucinations" that may lead AI Agents to take nonsensical or harmful actions; depending on the sensitivity activity they would be automating, these hallucinations could create significant harm to the end users they're supposed to represent, with significant detrimental impact on the trust end users would have in agents, but also possibly in online systems in general. The inability to establish a robust browser/Web distinction in LLM-based systems has already allowed prompt-injection attacks in many contexts. In an environment as open as the Web, and from a context as sensitive as the one managed today by browsers, such attacks could cause serious harm to users (see also AI in the Browser Threat Modeling).

A longstanding principle of Web architecture has been that the browser represents the user's interests. As the W3C Technical Architecture Group (TAG) writes, "A user agent acts as an intermediary between a person (its user) and the web... The user agent serves the person as a trustworthy agent: it always puts that person's interest first." As AI Agents transform the browser landscape, how will they earn our trust?

Browsers have (to a large extent) earned our user trust by performing "duties" on behalf of the user. Beyond any moral or legal considerations, these duties have played a critical role in enabling a flourishing ecosystem for both end users and content providers. The extent to which these duties can be applied in the less controlled environment induced by LLM-powered systems remains to be seen. For instance, browsers take pains to not expose the sites a user has already visited. Are there risks that an AI-based system would risk surfacing not only browsing history but maybe even prompt history?

Generally speaking, the current user experience of web browsers tells a relatively clear liability story: when something goes wrong we can tell whether there was user error, browser software error, or the problem lay with the Web site. When an AI agent takes more of the end user tasks and operates through a less deterministic model than a web browser, there is a significant risk of exposing users to an unreliable and untrustworthy platform. The risk will likely be compounded as the user's AI agents interact with a Web site's AI agents.

Another assumption that AI web user agents are proving fragile is the relatively clean delineation between end-user operated agents and crawlers. When crawlers emerged in the early days of the Web, the impact of their automated web page access on web servers led to the creation of the robots.txt proposal indicating which pages to crawl and limiting the frequency of access. In practice, there were only so many crawlers deployed at web scale that this proved sufficient in keeping the induced load on servers manageable. AI agents conducting deep research tasks for the end users can fetch many pages in quick succession in a completely distributed fashion across the population of end users (as painfully felt by a number of open source projects). Where robots.txt was meant to manage the impact of a few non-human directed crawlers, it is unlikely to be fit for purpose for widely distributed, user-directed agents (which may not be easily recognizable as traditional crawlers). Without a standard in place for managing these interactions, anti-AI firewalls have started to emerge, which creates friction and requires more resources. We see standardization opportunities to support responsible actors that wish to work together to reduce ecosystem costs.```
